File: src/Common/Sex.ml
type binary_handling = { as_male : bool; as_female : bool }

type t =
  | Male
  | Female
  | BalThani of binary_handling
  | Tsajana of binary_handling
  | Custom of { binary_handling : binary_handling; name : string }

module Decode = struct
  open Decoders_bs.Decode

  let binary_handling =
    field "as_male" bool
    >>= fun as_male ->
    field "as_female" bool >>= fun as_female -> succeed { as_male; as_female }

  let t =
    one_of
      [
        ( "Binary",
          string
          >>= function
          | "Male" -> succeed Male
          | "Female" -> succeed Female
          | _ -> fail "Expected a binary sex" );
        ( "Non-binary",
          field "type" string
          >>= function
          | "BalThani" ->
              field "value" (field "binary_handling" binary_handling)
              >|= fun opt -> BalThani opt
          | "Tsajana" ->
              field "value" (field "binary_handling" binary_handling)
              >|= fun opt -> Tsajana opt
          | "Custom" ->
              field "value"
                (field "binary_handling" binary_handling
                >>= fun binary_handling ->
                field "name" string
                >>= fun name -> succeed (Custom { binary_handling; name }))
          | _ -> fail "Expected a non-binary sex" );
      ]
end

module Encode = struct
  open Decoders_bs.Encode

  let binary_handling { as_male; as_female } =
    obj [ ("as_male", bool as_male); ("as_female", bool as_female) ]

  let t (sex : t) =
    match sex with
    | Male -> string "Male"
    | Female -> string "Female"
    | BalThani bin ->
        obj
          [
            ("type", string "BalThani");
            ("value", obj [ ("binary_handling", binary_handling bin) ]);
          ]
    | Tsajana bin ->
        obj
          [
            ("type", string "BalThani");
            ("value", obj [ ("binary_handling", binary_handling bin) ]);
          ]
    | Custom { binary_handling = bin; name } ->
        obj
          [
            ("type", string "BalThani");
            ( "value",
              obj
                [
                  ("binary_handling", binary_handling bin); ("name", string name);
                ] );
          ]
end

File: src/shared/domain/sex.ts
import { BinarySex as RawBinarySex } from "optolith-database-schema/types/_Sex"
import { SexualCharacteristic } from "optolith-database-schema/types/prerequisites/single/SexualCharacteristicPrerequisite"
import { assertExhaustive } from "../utils/typeSafety.ts"

/**
 * The character's sex. It does not have to be binary, although it always must be specified how to handle it in the context of binary sex prerequisites. You can also provide a custom sex with a custom name.
 */
export type Sex = BinarySex | NonBinarySex | CustomSex

/**
 * A binary sex option.
 */
export type BinarySex = {
  type: "Male" | "Female"
}

/**
 * A non-binary sex option.
 */
export type NonBinarySex = {
  type: "BalThani" | "Tsajana"

  /**
   * Defines how a non-binary sex should be treated when checking prerequisites.
   */
  ruleBasedInterpretation: RuleBasedInterpretation
}

/**
 * A custom non-binary sex option.
 */
export type CustomSex = {
  type: "Custom"

  /**
   * The custom sex name.
   */
  name: string

  /**
   * Defines how a non-binary sex should be treated when checking prerequisites.
   */
  ruleBasedInterpretation: RuleBasedInterpretation
}

/**
 * Defines how a non-binary sex should be treated when checking prerequisites.
 */
export type RuleBasedInterpretation = {
  /**
   * Defines if the sex should be treated as male when checking prerequisites.
   */
  asMale: boolean

  /**
   * Defines if the sex should be treated as female when checking prerequisites.
   */
  asFemale: boolean

  /**
   * Defines how a non-binary sex’s sexual characteristics should be treated
   * when checking prerequisites.
   */
  sexualCharacteristics: SexualCharacteristicsInterpretation
}

/**
 * Defines how a non-binary sex’s sexual characteristics should be treated when
 * checking prerequisites.
 */
export type SexualCharacteristicsInterpretation = {
  /**
   * Defines if the sex should be treated as having a penis when checking
   * prerequisites.
   */
  hasPenis: boolean

  /**
   * Defines if the sex should be treated as having a vagina when checking
   * prerequisites.
   */
  hasVagina: boolean
}

/**
 * Matches a binary sex from the database against the actual sex of the
 * character. It returns `true` if the actual sex either counts as the binary
 * sex by default or the user has chosen to, `false` otherwise.
 */
export const matchBinarySexToSex = (sex: Sex, binarySex: RawBinarySex): boolean => {
  switch (binarySex) {
    case "Male": {
      switch (sex.type) {
        case "Male":
          return true
        case "Female":
          return false
        case "BalThani":
        case "Tsajana":
        case "Custom":
          return sex.ruleBasedInterpretation.asMale
        default:
          return assertExhaustive(sex)
      }
    }
    case "Female": {
      switch (sex.type) {
        case "Male":
          return false
        case "Female":
          return true
        case "BalThani":
        case "Tsajana":
        case "Custom":
          return sex.ruleBasedInterpretation.asFemale
        default:
          return assertExhaustive(sex)
      }
    }
    default:
      return assertExhaustive(binarySex)
  }
}

/**
 * Matches a sexual characteristic from the database against the sex of the
 * character. It returns `true` if the actual sex either counts as having the
 * sexual characteristic by default or the user has chosen to, `false`
 * otherwise.
 */
export const matchSexualCharacteristicToSex = (
  sex: Sex,
  sexualCharacteristic: SexualCharacteristic,
): boolean => {
  switch (sexualCharacteristic) {
    case "Penis": {
      switch (sex.type) {
        case "Male":
          return true
        case "Female":
          return false
        case "BalThani":
        case "Tsajana":
        case "Custom":
          return sex.ruleBasedInterpretation.sexualCharacteristics.hasPenis
        default:
          return assertExhaustive(sex)
      }
    }
    case "Vagina": {
      switch (sex.type) {
        case "Male":
          return false
        case "Female":
          return true
        case "BalThani":
        case "Tsajana":
        case "Custom":
          return sex.ruleBasedInterpretation.sexualCharacteristics.hasVagina
        default:
          return assertExhaustive(sex)
      }
    }
    default:
      return assertExhaustive(sexualCharacteristic)
  }
}
